04_condition_variable/exercises/03b_ping_pong_easier.cpp

```diff
@@ -12,67 +13,90 @@ class PingPong {
     bool isPingTurn_ = true;
     int repetitions_;
     atomic<bool> play_{true};
+    int repsPing_ = 0;
+    int repsPong_ = 0;
 
 public:
     PingPong(int repetitions)
         : repetitions_(repetitions)
     {}
 
     void ping() {
-        thread_local int reps = 0;
-        while (reps < repetitions_ and play_)
+        while (repsPing_ <= repetitions_ and play_)
         {
-            unique_lock<mutex> l(m_);
-            // TODO: wait to be used here + printing, reps incrementation, chainging turn to pong
-            opponentsTurn_.wait(l, [&](){
-                return isPingTurn_;
-            });
-            std::cout << "ping " <<reps+1 <<std::endl;
-            reps++;
-            isPingTurn_ = false;
-            l.unlock();
+            if(not play_){
+                break;
+            }
+            
+            if(repsPing_ < repetitions_){
+                unique_lock<mutex> l(m_);
+                opponentsTurn_.wait(l, [&](){
+                    return (isPingTurn_ and (repsPing_ == repsPong_));
+                });
+                printingFunction("ping", (repsPing_+1));
+                repsPing_++;
+                isPingTurn_ = false;
+                this_thread::sleep_for(500ms);
+            }
+
+            else
+                printingFunction("Ping repetition limit");
+
+            opponentsTurn_.notify_all();
             this_thread::sleep_for(500ms);
         }
-        if (reps >= repetitions_)
-        {
-            // TOOD: only print message here
-            std::cout << "Ping repetition limit." << std::endl;
-        }
     }
 
     void pong() {
-        thread_local int reps = 0;
-        while (reps < repetitions_ and play_)
+        while (repsPong_ <= repetitions_ and play_)
         {
-            unique_lock<mutex> l(m_);
-            // TODO: wait to be used here + printing, reps incrementation, chainging turn to ping
-            opponentsTurn_.wait(l, [&](){
-                return not isPingTurn_;
-            });
-            std::cout << "pong " <<reps+1 << std::endl;
-            reps++;
-            isPingTurn_ = true;
-            l.unlock();
-            this_thread::sleep_for(500ms);
-        }
-        if (reps >= repetitions_) {
-            // TODO:  set play_ to false, display message, notify others to avoid deadlocks
-            play_ = false;
-            std::cout << "Pong repetition limit. \nEnd of program" << std::endl;
+            if(not play_){
+                break;
+            }
+            if(repsPong_ < repetitions_){
+                unique_lock<mutex> l(m_);
+                opponentsTurn_.wait(l, [&](){
+                    return ((not isPingTurn_) and(repsPong_ == repsPing_ -1));
+                });
+
+                printingFunction("pong", (repsPong_+1));
+                repsPong_++;
+                isPingTurn_ = true;             
+            }
+
+            else{
+                printingFunction("Pong repetition limit.\nEnd of program");
```

Comment:
to wypisywanie nie jest pod mutexem. Unique_lock chroni tylko tę część w ifie. Stąd tutaj dostaję wyścig.